Added option to use Virtualenv python version instead Pyenv global one

0x3333 opened the pull request:
#### Description

Currently, when you use `Pyenv` to manage python versions with `Virtualenv`, the prompt shows the `Pyenv` version, not the `Virtualenv` one, which is confusing.

This PR adds an option, `SPACESHIP_PYENV_USE_VENV`, which will enable a check if we are inside a `Virtualenv` and use its version instead `Pyenv`. This options defaults to `false`, which will not break any current setup.

This PR is related to PR #431.

#### Screenshot

![screen shot 2018-10-24 at 11 21 11](https://user-images.githubusercontent.com/289289/47440141-f8396a00-d783-11e8-99c7-c85d83207505.png)

salmanulfarzy commented:
Thank you for the PR @0x3333. 

I would like to clear somethings from my perspective. This would be the expected behavior, While `Pipenv` uses binaries from `pyenv` to create virtual environments, It doesn't modify the values in `pyenv` and `pyenv` does't handle virtual environments. So still the version known by `pyenv` is the one set by it and not the one in current environment. Including version from virtual env into `pyenv` section breaks the rule of least surprise. 

Honestly I don't much use `pyenv` and `venv` does the job for me. Both `venv` and `pyenv` sections serves different needs and users.

salmanulfarzy commented on sections/pyenv.zsh:
```diff
@@ -13,21 +13,27 @@ SPACESHIP_PYENV_PREFIX="${SPACESHIP_PYENV_PREFIX="$SPACESHIP_PROMPT_DEFAULT_PREF
   # Show pyenv python version only for Python-specific folders
   [[ -f requirements.txt ]] || [[ -n *.py(#qN^/) ]] || return
 
-  spaceship::exists pyenv || return # Do nothing if pyenv is not installed
```
Please don't make changes irrelevant to the PR. This flow of conditionals are styles followed across code base.

0x3333 replied:
Right, my bad, I thought that this check after checking for python files was an error. I'll fix.

salmanulfarzy commented:
Thank you for the PR @0x3333. 

Isn't this the expected behavior ? The version known by `pyenv` is the one set by it and not the one in current environment. Both `venv` and `pyenv` sections serves different needs and users. Including version from virtual env into `pyenv` section is against the rule of least surprise.

0x3333 commented:
Well, regarding the expected behavior, I'd say that yes. `pyenv` section should care about `pyenv` not `venv`.

But I believe that current behavior is kind of weird.

If you are working in a `virtualenv` you should see the python version in this `env` not the default one, which by the way, could not be invoke in the `venv`.

I don't know if this is the best solution, but we could think about how to "fix" this issue.

I'm open to alternatives to implement.

salmanulfarzy commented:
> But I believe that current behavior is kind of weird.

Agree on that and I don't use `pyenv` section.

> We could "care" about `pyenv`/`virtualenv`/`pipenv` and more inside it instead having separated sections.

Most of the `python` projects are inside virtual environments and `venv` section handles them and some in `conda` section. `Pipenv` is addressed with `venv` section I believe. 

May this is just me, unless `python` versions are globally switched frequently on dev setups, I don't see the purpose of `pyenv` section in prompt. It was introduced pretty early with #44 and #46, Both of them doesn't present any arguments. I'm aware about the use of `pyenv` by `Pipenv` for run time resolution, But it doesn't need to be shown on prompt. 

// @SanketDG @tienthanh8490 @Deckon @JonyEpsilon

0x3333 commented:
> Most of the `python` projects are inside virtual environments and `venv` section handles them and some in `conda` section. `Pipenv` is addressed with `venv` section I believe.

Yep, `pipenv` is addressed by `venv` perfectly.

> May this is just me, unless `python` versions are globally switched frequently on dev setups, I don't see the purpose of `pyenv` section in prompt.

I switch python version very often, some libs/progs are not compatible with some versions or I'm working in something else..

> I'm aware about the use of `pyenv` by `Pipenv` for run time resolution, But it doesn't need to be shown on prompt.

You are right, I think that the issue here is using `pyenv` to show python version, and not `pyenv` itself.

My use case is to show the python version in use, inside/outside a `venv`.

The problem is that if we remove `pyenv` and add this option to `venv`, I'll lost the ability to see the python version in a folder with python files outside a `venv`.

salmanulfarzy commented:
> Also, we could remove `pyenv` and create a `python`, which will show current python version(Invoking `python` instead `pyenv`), it would work inside and outside a `venv`.

This breaks existing behavior but sounds good. Hopefully we could address this with v4 release.

0x3333 commented:
Well, I did implement the Python section and reverted `pyenv`.

I also changed `pyenv` to `python` as default in the documentation. Let me know if this is OK.

Also, I added a check for `Pipfile` the new `requirements.txt` for Python.

If you need more work on this just let me know. If it can't be up-streamed you can close this PR.

![screen shot 2018-10-25 at 08 26 26](https://user-images.githubusercontent.com/289289/47498208-fed2ea80-d832-11e8-9869-a06f2fc8bf80.png)

0x3333 commented:
Well, I don't know if it should replace `pyenv` because it would cause breaking changes. That's why you see `pyenv` replaced by `python` in the documentation, this is not a leftover, I made `python` default, but did not removed `pyenv`.

The problem is that when inside a `venv` you can have a different python version than the global/shell version, and using `pyenv` section with `venv` you have a inconsistent information. See this screenshot: [Issue 431 comment](https://github.com/denysdovhan/spaceship-prompt/pull/431#issuecomment-432665068)

In my screenshot, you can see that the global/shell is `3.7.0` but in the `venv` it is `3.6.6`, this version shows the "correct" version of python when inside a `venv`.

salmanulfarzy commented:
With sections we either completely remove it or disable it with `_SHOW` variable. This PR was in the middle of that, `pyenv` was removed from prompt loading but options were kept in docs. Got me confused with the direction of this PR. 

I was thinking about merging these sections into one and dropping `pyenv` section in favor of `python` section. This will only have one section regarding `python` version.

|Variable|Value|Meaning|
|:----:|:--:|:---|
| `SPACESHIP_PYTHON_SHOW` | always | Always display python version information. |
| `SPACESHIP_PYTHON_SHOW` | true (default) | Display python version only inside virtual environment or if either `.python-version` or `pyenv-version` file exists | 
| `SPACESHIP_PYTHON_SHOW` | false | Don't display python version in prompt | 

With v4 we could deprecate `pyenv` and load `python`, Which would be acceptable with major release.

0x3333 commented:
I see, looks good to me. What do you need me to do? Remove completely pyenv?
Is there a branch for 4.0 version or all the work in done in the master?

salmanulfarzy commented:
I'll add this to `v4.0.0` milestone and it would be better if we wait for opinions from others about the suggested depreciation.

avarun42 commented:
Is this closed by #431?

salmanulfarzy commented:
> Is this closed by #431

Yes. `python` section would show `python` version inside virtual environment (_optionally_ `always`) regardless of how it got there. It'll be available with our next release. Thank you @0x3333 for contributing.
